File: Objective-C++/ArrayList/ArrayList.m
//
//  ArrayList.m
//  ArrayList
//
//  Created by sw on 2019/4/17.
//  Copyright © 2019 WS. All rights reserved.
//

#import "ArrayList.h"
#import "Assert.h"

#define DEFAULT_SIZE 10
#define ELEMENT_NOT_FOUND -1

@interface ArrayList ()
@property(nonatomic) NSUInteger size;
@property(nonatomic) NSMutableArray *elements;
//NSUInteger _size;
//NSMUtableArray *_elements;
@end

@implementation ArrayList
+ (instancetype)arraylist {
    return [[ArrayList alloc] init];
}

-(instancetype)init {
    if (self = [super init]) {
        _elements = [NSMutableArray arrayWithCapacity:DEFAULT_SIZE];
        _size = 0;
    }
    return self;
}

- (void)add:(NSObject *)obj {
    [Assert assertNil:obj];
    _elements[_size++] = obj;
}

- (void)pop {
    [_elements removeLastObject];
    _size -= 1;
}

- (void)removeObject:(NSObject *)obj {
    [Assert assertNil:obj];
    if ([self contains:obj] == NO) {
        return;
    }
    NSInteger idx = [self indexOfObject:obj];
    [self removeObjectAtIndex:idx];
//    for (NSInteger i = idx; i < _size - 1; i++) {
//        _elements[i] = _elements[i+1];
//    }
//    [_elements removeLastObject];
//    _size -= 1;
}

- (void)removeObjectAtIndex:(NSInteger)idx {
    [Assert assertArraySize:_size idx:idx];
    for (NSInteger i = idx; i < _size - 1; i++) {
        _elements[i] = _elements[i+1];
    }
    // _elements[_size - 1] = nil; // 置空会crash
    [_elements removeLastObject];
    _size -= 1;
}

- (void)insert:(NSObject *)obj atIndex:(NSInteger)idx {
    [Assert assertNil:obj];
    [Assert assertArraySize:_size idx:idx];
    
    for (NSInteger i = _size; i > idx; i--) { // idx后面所有的元素向后移动一位
        _elements[i] = _elements[i - 1];
    }
    _elements[idx] = obj;
    _size += 1;
}

- (NSObject *)objectAtIndex:(NSInteger)idx {
    [Assert assertArraySize:_size idx:idx];
    return _elements[idx];
}

- (NSInteger)indexOfObject:(NSObject *)obj {
    [Assert assertNil:obj];
    for (NSInteger i = 0; i < _size; i++) {
        if ([_elements[i] isEqual:obj]) {
            return i;
        }
    }
    return ELEMENT_NOT_FOUND;
}

- (NSObject *)setObject:(NSObject *)obj atIndex:(NSInteger)idx {
    [Assert assertNil:obj];
    [Assert assertArraySize:_size idx:idx];
    NSObject *oldElement = _elements[idx];
    _elements[idx] = obj;
    return oldElement;
}

- (BOOL)contains:(NSObject *)obj {
    [Assert assertNil:obj];    
    return [self indexOfObject:obj] != ELEMENT_NOT_FOUND;
}

- (void)clear {
    _elements = [NSMutableArray arrayWithCapacity:DEFAULT_SIZE];
    _size = 0;
}

- (BOOL)isEmpty {
    return _size == 0;
}

- (NSUInteger)size {
    return _size;
}

- (NSString *)description {
    if ([self isEmpty] == YES) {
        return [super description];
    }
    NSString *desc = _elements[0];
    for (int i = 1; i < _size; i++) {
        desc = [NSString stringWithFormat:@"%@,%@",desc, _elements[i]];
    }
    return desc;
}

#pragma mark - Private
- (BOOL)checkNil:(NSObject *)obj {
    if (obj == nil) {
        return YES;
    }
    return NO;
}

- (BOOL)checkRangeException:(NSInteger)idx {
    if (idx < 0 || idx >= _size) {
        return YES;
    }
    return NO;
}
@end

File: Python/List/ArrayList.py
import os


class ArrayList:
    ```创建一个arraylist对象```

    # size
    # defaultCount
    # elements
    # 对象方法、类方法、静态方法、对象属性、类属性前加两个下划线代表private私有

    __DEFAULT_COUNT = 10
    # 限制对象只可以动态增加这两个对象属性
    __slots__ = ['__elements', '__size']  # 对象属性前加两个下划线代表私有，外界不可访问

    def __init__(self):
        self.__elements = []
        self.__size = 0
        # print('init __size = %d'%(self.__size))

    # 类方法
    @classmethod
    def array_list(cls):
        return cls()  # 也可以 return ArrayList()

    # 对象方法
    def add(self, obj):
        self.__elements.append(obj)  # python 中不能使用list[idx] = xxx来添加元素，否则会报越界错误
        self.__size += 1
        print(self.__elements, self.__size)

    def remove_at_index(self, idx):
        for i in range(idx, self.__size - 1):
            print(i)
            self.__elements[idx] = self.__elements[idx + 1]
        del self.__elements[self.__size - 1]  # 目前只想到这个方法，python貌似不支持设置为null
        # self.__elements[self.__size - 1] = None  # 不能使用None,None也是对象
        self.__size -= 1
        print(self.__elements)

    def remove(self, obj):
        idx = self.index_of(obj)  # 之所以不用contains方法，因为避免先调用contains、再调用index_of而消耗性能
        if idx == -1:
            return
        self.remove_at_index(idx)

    def pop(self):
        del self.__elements[self.__size - 1]
        # self.__elements[self.__size - 1] = None
        self.__size -= 1

    def object_at_index(self, idx):
        return self.__elements[idx]

    def index_of(self, obj):
        for i in range(0, self.__size):
            if self.__elements[i] == obj:
                return i
        return -1

    def contains(self, obj):
        return self.index_of(obj) != -1

    def is_empty(self):
        return self.__size == 0

    def clear(self):
        self.__elements = []

    def size(self):
        return self.__size


# list = ArrayList()
# list.add(12)
# print(list.elements)
# list.elements = [2, 4]
# print(list.elements)

list = ArrayList.array_list()
# list.__elements = [999]
# print(list.__elements)

list.add(12)
print(list)
list.add(24)
list.remove_at_index(0)
list.add(666)
list.add(888)

print(list.index_of(666))
print(list.object_at_index(2))
print(list.contains(24))
print (list.is_empty())
